contracts/tokenomics/generator/src/contract.rs

```diff
@@ -1390,32 +1463,251 @@ fn send_orphan_proxy_rewards(
+            admin: Some(owner.to_string()),
+            code_id: whitelist_code_id,
             funds: vec![],
+            label: "Proxy rewards holder".to_string(),
+            msg: to_binary(&astroport::whitelist::InstantiateMsg {
+                admins: vec![admin.to_string()],
+                mutable: false,
+            })?,
+        }),
+        INIT_REWARDS_HOLDER_ID,
+    );
+
+    Ok(msg)
+}
+
+fn migrate_proxy(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    lp_token: String,
+    new_proxy: String,
+) -> Result<Response, ContractError> {
+    let lp_addr = addr_validate_to_lower(deps.api, &lp_token)?;
+    let new_proxy_addr = addr_validate_to_lower(deps.api, &new_proxy)?;
+
+    let cfg = CONFIG.load(deps.storage)?;
+
+    // Permission check
+    if info.sender != cfg.owner {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    if !cfg.allowed_reward_proxies.contains(&new_proxy_addr) {
+        return Err(ContractError::RewardProxyNotAllowed {});
+    }
+
+    // Check the pool has reward proxy
+    let pool_info = POOL_INFO.load(deps.storage, &lp_addr)?;
+    if let Some(proxy) = &pool_info.reward_proxy {
+        if proxy.as_str() == new_proxy_addr.as_str() {
+            return Err(StdError::generic_err("Can not migrate to the same proxy").into());
+        }
+    } else {
+        return Err(StdError::generic_err("Pool does not have proxy").into());
+    }
+
+    update_rewards_and_execute(
+        deps,
+        env,
+        Some(lp_addr.clone()),
+        ExecuteOnReply::MigrateProxy {
+            lp_addr,
+            new_proxy_addr,
+        },
+    )
+}
+
+fn migrate_proxy_callback(
+    mut deps: DepsMut,
+    env: Env,
+    lp_addr: Addr,
+    new_proxy_addr: Addr,
+) -> Result<Response, ContractError> {
+    let mut pool_info = POOL_INFO.load(deps.storage, &lp_addr)?;
+    let cfg = CONFIG.load(deps.storage)?;
+    accumulate_rewards_per_share(&deps.querier, &env, &lp_addr, &mut pool_info, &cfg, None)?;
+
+    // We've checked this before the callback so it's safe to unwrap here
+    let prev_proxy_addr = pool_info.reward_proxy.clone().unwrap();
+
+    let proxy_lp_balance: Uint128 = deps
+        .querier
+        .query_wasm_smart(&prev_proxy_addr, &ProxyQueryMsg::Deposit {})?;
+
+    // Since we migrate to another proxy the proxy reward balance becomes 0.
+    pool_info.proxy_reward_balance_before_update = Uint128::zero();
+    // Save a new index and orphan rewards for the new proxy
+    pool_info
+        .accumulated_proxy_rewards_per_share
+        .update(&new_proxy_addr, Decimal::zero())?;
+    pool_info
+        .orphan_proxy_rewards
+        .update(&new_proxy_addr, Uint128::zero())?;
+    // Set new proxy
+    pool_info.reward_proxy = Some(new_proxy_addr.clone());
+
+    POOL_INFO.save(deps.storage, &lp_addr, &pool_info)?;
+
+    update_proxy_asset(deps.branch(), &prev_proxy_addr)?;
+    update_proxy_asset(deps.branch(), &new_proxy_addr)?;
```

Comment:
Maybe update only for new one and for old do it on migration?

Reply:
Yep, it could be better. Fixed